### ISSUE TITLE
hvd.size() is 1 but MPI.COMM_WORLD.Get_size() is 8 

### ISSUE THREAD

**Issue by TenTen-Teng** [horovod/horovod#2588]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet): TensorFlow
2. Framework version: tf2.3.0
3. Horovod version: 0.21.0
4. MPI version: 4.1.0
5. CUDA version: 10.1
6. NCCL version: 2.7.8
7. Python version: 3.8.7
8. OS and version: ubuntu 18.04
9. GCC version: 7.5.0

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
https://github.com/horovod/horovod/issues/1646

**Your question:**
I am trying to run the MNIST example using Horovod. I'd like to run the example with 8 GPUs, but only one GPU was activated during training. I printout `hvd.rank()`, `hvd.local_rank()`, `hvd.size()`, `MPI.COMM_WORLD.Get_size()` and `MPI.COMM_WORLD.Get_rank()`. It shows that `hvd.size()` is always 1 and `hvd.local_rank()`/`hvd.rank()` is always 0, but `MPI.COMM_WORLD.Get_size()` is 8 and `MPI.COMM_WORLD.Get_rank()` is from 0 - 7. 

In my code,  I pin GPU to be used to process local rank by using
```python3
hvd.init()

gpus = tf.config.experimental.list_physical_devices("GPU")
for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)
if gpus:
    tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], "GPU")
```

The shell script I use to call python script:
```
mpirun -np 8 \
    --hostfile ${PBS_NODEFILE} \
    -bind-to none \
    -map-by slot \
    -mca btl_tcp_if_exclude lo, docker0,eth0 \
    -mca pml ob1 \
    -mca btl ^openib \
    -x NCCL_DEBUG=INFO \
    -x LD_LIBRARY_PATH \
    -x PATH \
    -x PYTHONPATH \
    python3 tf_keras.py 
```

However, if I call the python script using `horovodrun`, all 8 gpus are working properly. The `hvd.size()` returns 8 and the rank is 0-7, but `MPI.COMM_WORLD.Get_size()` is always 1 this time. 
```
horovodrun -np 8 -H localhost:8 python3 tf_keras.py
```

A little more test, instead of using `hvd.size()` and `hvd.rank()`, I use `MPI.COMM_WORLD.Get_rank()` to get rank and `MPI.COMM_WORLD.Get_size()` to get world size in my code. In this case, if I run the script with `mpirun` command, the 8 GPUs work properly. However, if I run the script with `horovodrun` command, then only 1 GPU was activated.  The printout results are just the opposite of last time.
```python3
hvd.init()

gpus = tf.config.experimental.list_physical_devices("GPU")
for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)
if gpus:
    tf.config.experimental.set_visible_devices(gpus[MPI.COMM_WORLD.Get_rank()], "GPU")
```

It seems like if I write my code with `hvd.size`, I have to call it with `horovodrun`.  If I use `mpirun` to call it, there will be only 1 GPU activated. The same thing happens if I use `MPI.COMM_WORLD.Get_size()` but run it use `horovodrun`. However, I expect to use either of the commands to run the script. They should return the same rank number...

**Comment by TenTen-Teng** [comment]
I also wrote a python script only printout `hvd.size` and `comm.size` for testing:
```python3
import horovod.tensorflow as hvd
from mpi4py import MPI

comm = MPI.COMM_WORLD

hvd.init()

while 1:
    print("****** hvd rank is: %s" % hvd.rank())
    print("****** hvd local_rank is: %s" % hvd.local_rank())
    print("****** hvd world size is: %s" % hvd.size())
    print("****** COMM world size is: %s" % comm.Get_size())
    print("****** COMM rank is: %s" % comm.Get_rank())
```
I tried to run it with horovod 0.19.0 and 0.21.0
With horovod 0.19.0 + mpi 3.0.3, `hvd.size` and `comm.size` shows both 8
With horovod 0.21.0 + mpi 4.1.0:
    - call it with `horovodrun`, `hvd.size` is 8 and ` comm.size` is always 1
    - call it with `mpirun`, `hvd.size` is always 1 and `comm.size` is 8

Is there any packages mismatching problem (like horovod & mpi?)

**Comment by tgaddair** [comment]
Hey @TenTen-Teng, can you run `horovodrun --check-build` and share the results?

It could be that you have more than one version of MPI installed, and Horovod is picking up the wrong version when it initializes the communicator.

**Comment by TenTen-Teng** [comment]
I built two docker images to run my code, one with horovod 0.19.0 and the other with horovod 0.21.0.
The results from horovod 0.21.0:
```
2021-01-12 20:45:46.201878: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-12 20:45:47.968639: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-12 20:45:49.547072: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-12 20:45:51.144570: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-12 20:45:52.777683: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-12 20:45:54.392220: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
Horovod v0.21.0:

Available Frameworks:
    [X] TensorFlow
    [ ] PyTorch
    [ ] MXNet

Available Controllers:
    [X] MPI
    [X] Gloo

Available Tensor Operations:
    [ ] NCCL
    [ ] DDL
    [ ] CCL
    [X] MPI
    [X] Gloo
```

The results from horovod 0.19.0: 
```
2021-01-12 20:40:11.787634: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-12 20:40:13.657539: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-12 20:40:15.554517: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-12 20:40:17.417605: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-12 20:40:19.291371: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-12 20:40:21.168173: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
Horovod v0.19.0:

Available Frameworks:
    [X] TensorFlow
    [ ] PyTorch
    [ ] MXNet

Available Controllers:
    [X] MPI
    [X] Gloo

Available Tensor Operations:
    [ ] NCCL
    [ ] DDL
    [ ] CCL
    [X] MPI
    [X] Gloo 
```

**Comment by chongxiaoc** [comment]
Not a problem to me using your above script, build horovod from master:

```
(hvd-py3.7) chongxiaoc@chongxiaoc-C02DL7WPMD6R python_sandbox % horovodrun -np 4 python hvd_info.py
[chongxiaoc-C02DL7WPMD6R:18103] Warning: could not find environment variable "PS1"
[1,1]<stdout>:****** hvd rank is: 1
[1,1]<stdout>:****** hvd local_rank is: 1
[1,1]<stdout>:****** hvd world size is: 4
[1,1]<stdout>:****** COMM world size is: 4
[1,1]<stdout>:****** COMM rank is: 1
[1,2]<stdout>:****** hvd rank is: 2
[1,2]<stdout>:****** hvd local_rank is: 2
[1,2]<stdout>:****** hvd world size is: 4
[1,2]<stdout>:****** COMM world size is: 4
[1,2]<stdout>:****** COMM rank is: 2
[1,0]<stdout>:****** hvd rank is: 0
[1,3]<stdout>:****** hvd rank is: 3
[1,3]<stdout>:****** hvd local_rank is: 3
[1,3]<stdout>:****** hvd world size is: 4
[1,3]<stdout>:****** COMM world size is: 4
[1,0]<stdout>:****** hvd local_rank is: 0
[1,0]<stdout>:****** hvd world size is: 4
[1,0]<stdout>:****** COMM world size is: 4
[1,0]<stdout>:****** COMM rank is: 0
[1,3]<stdout>:****** COMM rank is: 3

(hvd-py3.7) chongxiaoc@chongxiaoc-C02DL7WPMD6R python_sandbox % mpirun -np 4 python hvd_info.py
****** hvd rank is: 1
****** hvd local_rank is: 1
****** hvd world size is: 4
****** COMM world size is: 4
****** COMM rank is: 1
****** hvd rank is: 3
****** hvd local_rank is: 3
****** hvd world size is: 4
****** COMM world size is: 4
****** COMM rank is: 3
****** hvd rank is: 0
****** hvd local_rank is: 0
****** hvd world size is: 4
****** COMM world size is: 4
****** COMM rank is: 0
****** hvd rank is: 2
****** hvd local_rank is: 2
****** hvd world size is: 4
****** COMM world size is: 4
****** COMM rank is: 2
```

**Comment by TenTen-Teng** [comment]
Hi, @chongxiaoc thanks for running the script. I solved my problem. I rebuilt the image followed [this](https://github.com/horovod/horovod/blob/master/Dockerfile.gpu) with horovod 0.21.1 and openmpi 4.0.0. I could see those results as well now. I guess something wrong with my dockerfile. I installed horovod using `pip3 install horovod` in my old dockerfile. I'd better double check the dockerfile. Thanks, @tgaddair, for your time to help me with this.
